docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,253 @@
+   scope of this proposal.
+ - adding support to Windows containers. Windows container support is a
+   parallel effort and is tracked by [#22623](https://issues.k8s.io/22623).
+   The new interface will not be augmented to support Windows containers, but
+   it will be made extensible such that the support can be added in the future.
+ - re-defining Kubelet's internal interfaces. These interfaces, though, may
+   affect Kubelet's maintainability, is not relevant to runtime integration.
+ - improving Kubelet's efficiency or performance, e.g., adopting event stream
+   from the container runtime [#8756](https://issues.k8s.io/8756),
+   [#16831](https://issues.k8s.io/16831).
+
+## Requirements
+
+ * Support existing integrated container runtimes: `docker` and `rkt`.
+ * Support hypervisor-based container runtimes: `hyper`.
+
+## Container Runtime Interface
+
+The main idea of the proposal is to adopt an imperative container-level
+interface, which allows Kubelet to control the lifecycles of the containers
+inside a pod, across all container runtimes. A separate PodSandbox concept
+is defined to represent the environment in which a group of containers in a
+pod will be created and run. The container runtimes may interpret the sandbox
+concept differently based on how it operates internally. For runtimes relying
+on hypervisor, sandbox represents a virtual machine. For others, it can be
+Linux namespaces.
+
+In short, a PodSandbox should have the following features.
+
+ * **Isolation**: E.g., Linux namespaces or a full virtual machine, or even
+   support additional security features.
+ * **Compute resource Specifications**: A PodSandbox should implement pod-level
+   resource demands and restrictions.
+
+A container in a PodSandbox maps to a application in the Pod Spec. For Linux
+containers, they are expected to share at least network and IPC namespaces,
+with sharing more namespaces discussed in [#1615](https://issues.k8s.io/1615).
+
+
+Below is an example of the proposed interfaces.
+
+```go
+// PodSandboxManager contains basic operations for sandbox.
+type PodSandboxManager interface {
+	Create(config *PodSandboxConfig) (string, error)
+	Delete(id string) (string, error)
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	Status(id string) PodSandboxStatus
+}
+
+// ContainerRuntime contains basic operations for containers.
+type ContainerRuntime interface {
+    Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, sandboxIDng) (string, error)
+    Start(id string) error
+    Stop(id string, timeout int) error
+    Remove(id string) error
+    List(filter ContainerFilter) ([]ContainerListItem, error)
+    Status(id string) (ContainerStatus, error)
+}
+
+// ContainerCommandExecutor provides methods to run commands in a container.
+type ContainerCommandExecutor interface {
+	Exec(id string, cmd []string, streamOpts StreamOptions) error
+}
+
+// ImageOperations contains image-related operations.
+type ImageOperations interface {
```

Comment:
This differs from the "go" file's version of the same, I think it would be very helpful to just have this link to the go file so there is a single canonical location for discussion.